Print final conflict set when search is exhaustive

fmthoma opened the pull request:
Pull request for #3570.

The final confilct set is printed along with the conflict count of each conflict, if available.

Open for discussion:
- The original error message is thrown away, we currently only show the conflilct set. Should we restore it and print both?
- The conflict count is not actually a useful metric. Should we expose it, or should we rescale it to something more useful to the user?

ezyang commented on cabal-install/Distribution/Solver/Modular/Log.hs:
```diff
@@ -40,25 +40,25 @@ logToProgress mbj l = let
     -- and ignores repeated backjumps. If proc reaches the backjump limit, it truncates
```
This comment needs updating.

ezyang commented on cabal-install/Distribution/Solver/Modular/Log.hs:
```diff
@@ -40,25 +40,25 @@ logToProgress mbj l = let
         replace _       (Done x)                          = Done x
-        replace cs'     (Fail cs)                         = -- 'Nothing' means backjump limit not reached.
+        replace cs'     (Fail (exh, cs))                  = -- 'Nothing' means backjump limit not reached.
                                                             -- Prefer first error over later error.
```
This comment needs updating (something about doing something specifically when you exhaust search space?)

ezyang commented on cabal-install/Distribution/Solver/Modular/Log.hs:
```diff
@@ -72,17 +72,16 @@ logToProgress mbj l = let
-                                                        currlimit Nothing  = ""
+    go ms _           (Fail (exh, cs))      = Fail $ "Could not resolve dependencies:\n" ++ if exh
+                                                then "Dependency tree exhaustively searched.\n" ++
+                                                         "Final conflict set is: " ++ CS.showCS cs
```
Not something you need to do for this patchset, but someday we should explain to mere mortals what a conflict set is :)

ezyang commented:
Since this is just user visible business, it doesn't look that easy to automated test. But it would be nice if the commit messages included some sample output, i.e., "What would I expect to see with this new change?" For example, something like in da388222533c52018fe94ad7951cb2b46540dec8

fmthoma commented:
@ezyang Thank you for your review comments! I have addressed them in my updates. Also I have changed the message to a slightly less technical »I've had most trouble fulfilling the following goals: …«.

ezyang commented:
Thanks! Really helpful. CI failure is real though; try building locally with -Werror.

I wonder if we should print the conflict counts in non-verbose mode. They don't seem that useful for someone who doesn't know what they are.

fmthoma commented:
@ezyang: I had this discussion with @kosmikus as well. On one hand, the absolute numbers do not provide any useful metric. On the other hand, their relative values do provide a tendency how severe a conflict is (there is a difference between `package1 (10000), package2 (9999)` and `package1 (10000), package2 (3)` that is not neglectable). So here are some suggestions:
* Print the numbers as they are
* Defer them to verbose mode
* Normalize them to something more useful (e.g. to percentage of all conflicts)
* Use them to further reduce the output, e.g. cut off all packages that contribute less than 1..10% of all conflicts.

The warning should be gone now, I overlooked a `Control.Applicative` import.

ezyang commented:
I'm OK with leaving the counts, I just want to open up a discussion.

Actually, while we're at it, why don't we talk about what the conflict counts mean in the first place. My understanding is they is they specify how many times a package shows up in a "conflict set", which is a set of packages that have incompatible constraints that cause a solution to be unsat. Is this accurate?  This makes me wonder: the solver is supposed to bail out early when it encounters a conflict. So the actual conflict counts are going to be heavily dependent on the order the solver attempted to solve packages in.

fmthoma commented:
> So the actual conflict counts are going to be heavily dependent on the order the solver attempted to solve packages in.

Yes, this shows up for example in #3961 where the counts are heavily dependent on how we accumulate them while backjumping. So these numbers are not an exact science, but they give a good indication nevertheless.

I personally would probably prefer from a user perspective to have percentages (»This package was involved in about n% of all conflicts«) and a sensible cutoff.

phadej commented:
@grayjay use

```hs
import Prelude ()         
import Distribution.Client.Compat.Prelude
```

(related to `Control.Applicative`)

grayjay commented on cabal-install/Distribution/Solver/Modular/Explore.hs:
```diff
@@ -67,8 +67,6 @@ backjump (EnableBackjumping enableBj) var initial xs =
 
```
I just realized that the line above updates the conflict map even when `countConflicts` is false.  It should probably be something like `if countConflicts then (cs, updateCM initial cm) else (cs, cm)` to avoid creating a thunk from the conflict map when it isn't used.

fmthoma replied:
Good catch, changed it in the PR.

grayjay replied:
@fmthoma Actually, the code that I linked to above isn't enough to prevent space leaks.  I had to fix it in https://github.com/haskell/cabal/pull/4122/commits/41cd78641926bfa05f9aa6d8e75e56be0f8ec18c.

`getFirstGoal` doesn't evaluate the conflict map unless there is more than one goal, so it is possible for a thunk to develop even with --count-conflicts.  Using `seq` is more reliable.

grayjay commented on cabal-install/Distribution/Solver/Modular/Log.hs:
```diff
@@ -20,11 +22,13 @@ import qualified Distribution.Solver.Modular.ConflictSet as CS
 messages :: Progress step fail done -> [step]
 messages = foldProgress (:) (const []) (const [])
 
+data Exhaustive = Exhaustive | NotExhaustive
```
I think `Exhaustiveness` would be clearer.

grayjay commented on cabal-install/Distribution/Solver/Modular/Log.hs:
```diff
@@ -39,26 +43,27 @@ logToProgress mbj l = let
         replace _       (Done x)                          = Done x
-        replace cs'     (Fail cs)                         = -- 'Nothing' means backjump limit not reached.
+        replace _       (Fail (Exhaustive,    cs, cm))    = Fail (Exhaustive, cs, cm)
+        replace cs'     (Fail (NotExhaustive, cs, cm))    = -- Backjump limit not reached.
```
Should this be "Backjump limit reached"?

fmthoma replied:
Indeed, I'll the `data Exhaustive` to `data Exhaustiveness = Exhaustive | BackjumpLimitReached` so the comment is now unnecessary.

grayjay commented on cabal-install/Distribution/Solver/Modular/Log.hs:
```diff
@@ -72,17 +77,21 @@ logToProgress mbj l = let
-                                                        currlimit Nothing  = ""
+    go ms _           (Fail (exh, cs, cm))  = Fail $
+        "Could not resolve dependencies:\n" ++
+        unlines (messages $ showMessages (L.foldr (\ v _ -> v `CS.member` cs) True) False ms) ++
```
It looks like `cs` is the first conflict set when the backjump limit is reached and the last conflict set when the search is exhaustive.  Shouldn't this line always filter the short log using the first conflict set?  The final conflict set might not contain the first error, which would cause the error to be filtered out.

Another option is to remove the filtered log altogether, to avoid showing two possibly unrelated errors.

grayjay replied:
The solver always prints `ms`, which is the log to the first error.  It truncates the log on the line `es = proc (Just 0) l -- catch first error (always)` above.

> BTW, what is the original rationale anyway for showing the first rather than the last error in case the backjump limit is reached?

According to the discussion in #941, the first error was chosen because the solver has made fewer compromises at that point, and the conflicts are more likely to be relevant.  Also, I don't think that there is an obvious place to start printing the summarized log leading up to the final error.  The log wouldn't always include the top of the tree.

fmthoma replied:
Okay, so to summarize:
* We always print the first error encountered.
* In case the backjump limit is reached, we print the first conflict set (which coincides with the first error).
* For an exhaustive search, we print the final conflict set (which may not coincide with the first error).

So this could be solved
1. either by choosing to print the first rather than the final conflict set also in the exhaustive case (what would this imply?),
2. or by printing the error only in the non-exhaustive case and thus avoiding the problem (in fact, in an early version of the PR, @kosmikus and I had printed the error message only in the non-exhaustive case, but we changed it),
3. or by printing the last error in the non-exhaustive case.

My understanding of the issue is still somewhat incomplete, but it seems to me that using the reasoning from #941 also applies to the exhaustive case, so option (1) looks to me like the soundest one?

grayjay replied:
I like option (2).

> 1. either by choosing to print the first rather than the final conflict set also in the exhaustive case (what would this imply?),

I think that the first conflict set isn't very useful, because it comes from the first node where the solver backjumps, which is near the leaves of the tree.  It contains the same information as the first error.  However, the final conflict set is assigned to the root of the tree (in the exhaustive case), and it summarizes all conflicts in the tree.  It means that any solution to the dependency problem would need to assign a value that isn't available in the search tree to at least one of the variables in the conflict set.

The solver's conflict sets differ from the ones in [Modular lazy search for Constraint Satisfaction Problems] (http://dx.doi.org/10.1017/S0956796801004051) because they can also contain variables that are below the current node.  That difference makes the final conflict set more useful.

> 3 . or by printing the last error in the non-exhaustive case.

I'm not sure how this makes the behavior more consistent, or did you mean the exhaustive case?

fmthoma replied:
Yes, I meant the exhaustive case :-) (fixed it in my comment above.)

Then let's go with option (2), I will do the change.

fmthoma replied:
Pushed the change. Unfortunately, the failing tests point out a problem with this approach: We don't know _why_ a package causes a problem (which might be trivial in case of a package conflict, but in the case of a missing unknown package, that information is quite valuable).

fmthoma replied:
How about a message conveying that the error does not need to be in line with the final conflict set:

```
> cabal install --dry-run cabal-install-1.24.0.0 cabal-install-1.22.9.0
Resolving dependencies...
cabal: Could not resolve dependencies:
next goal: cabal-install (user goal)
rejecting: cabal-install-1.24.0.0 (constraint from user target requires
==1.22.9.0)
rejecting: cabal-install-1.22.9.0 (constraint from user target requires
==1.24.0.0)
rejecting: cabal-install-1.22.8.0, cabal-install-1.22.7.0,
cabal-install-1.22.6.0, cabal-install-1.22.5.0, cabal-install-1.22.4.0,
cabal-install-1.22.3.0, cabal-install-1.22.2.0, cabal-install-1.22.0.1,
cabal-install-1.22.0.0, cabal-install-1.20.2.0, cabal-install-1.20.1.0,
cabal-install-1.20.0.6, cabal-install-1.20.0.5, cabal-install-1.20.0.4,
cabal-install-1.20.0.3, cabal-install-1.20.0.2, cabal-install-1.20.0.1,
cabal-install-1.20.0.0, cabal-install-1.18.2.0, cabal-install-1.18.1.0,
cabal-install-1.18.0.8, cabal-install-1.18.0.7, cabal-install-1.18.0.6,
cabal-install-1.18.0.5, cabal-install-1.18.0.4, cabal-install-1.18.0.3,
cabal-install-1.18.0.2, cabal-install-1.18.0.1, cabal-install-1.18.0,
cabal-install-1.16.1.0, cabal-install-1.16.0.2, cabal-install-1.16.0.1,
cabal-install-1.16.0, cabal-install-0.14.1, cabal-install-0.14.0,
cabal-install-0.10.2, cabal-install-0.10.0, cabal-install-0.8.2,
cabal-install-0.8.0, cabal-install-0.6.4, cabal-install-0.6.2,
cabal-install-0.6.0, cabal-install-0.5.2, cabal-install-0.5.1,
cabal-install-0.5.0, cabal-install-0.4.0 (constraint from user target requires
==1.22.9.0)
After searching the rest of the dependency exhaustively, these were the goals
I had most trouble fulfilling: cabal-install (49)
```

grayjay replied:
That looks good.  Should it be "the rest of the dependencies"?

grayjay replied:
That looks good.  So I think the only thing left is changing the summarized log so that it is always filtered with the first conflict set.

fmthoma replied:
Please have a look at the last commit. Now the summarized log should be filtered with the first conflict set. This involved removing `useFirstError` and changing the traversal of the log in the `go` function, making `go` partial again, so I hope I didn't break anything.

grayjay commented on cabal-install/Distribution/Solver/Modular/Log.hs:
```diff
@@ -72,17 +77,21 @@ logToProgress mbj l = let
+            Exhaustive ->
+                "Dependency tree exhaustively searched.\n" ++
+                "I've had most trouble fulfilling the following goals: "
+                ++ CS.showCSWithFrequency cm cs
```
I wonder if we should only show the counts at higher verbosity, since there are cases where they could be misleading.  For example, the solver could immediately stop after finding a missing dependency, which would give the missing dependency a count of 1, even though it is part of a minimal conflict set.

fmthoma commented:
@grayjay, @ezyang: I think it makes sense to show the conflict counts only on a higher verbosity level. However, this requires either passing the verbosity to the solver (which is kind of an intrusive change), or to change the return type of the solver (which is also quite intrusive). I am working on this, but I would rather suggest we merge this pull request without this change, and I open another one with the conflict counts only on verbose.
@kosmikus: Do you agree with @grayjay and @ezyang to show the counts only in verbose mode?

fmthoma commented:
@grayjay, @ezyang: I think it makes sense to show the conflict counts only on a higher verbosity level. However, this requires either passing the verbosity to the solver (which is kind of an intrusive change), or to change the return type of the solver (which is also quite intrusive). I am working on this, but I would rather suggest we merge this pull request without this change, and I open another one with the conflict counts only on verbose.
@kosmikus: Do you agree with @grayjay and @ezyang to show the counts only in verbose mode?

23Skidoo commented:
> I would rather suggest we merge this pull request without this change, and I open another one with the conflict counts only on verbose.

Fine with me.

grayjay commented on cabal-install/Distribution/Solver/Modular/Log.hs:
```diff
@@ -70,19 +60,29 @@ logToProgress mbj l = let
     --
     -- The third argument is the full log, ending with either the solution or the
     -- exhaustiveness and first conflict set.
```
This is now the final conflict set.

grayjay commented on cabal-install/Distribution/Solver/Modular/Log.hs:
```diff
@@ -20,48 +21,37 @@ import qualified Distribution.Solver.Modular.ConflictSet as CS
-                                                            Fail (isNothing cs, cs' <|> cs)
-        replace Nothing (Step x@(Failure cs Backjump) xs) = Step x $ replace (Just cs) xs
-        replace cs'     (Step x                       xs) = Step x $ replace cs' xs
-
```
Nice!

23Skidoo commented:
@grayjay @fmthoma Feel free to merge once the CI is green if you think that this PR is ready to go in.

grayjay commented:
It's green, so I'll merge it.  Thanks, @fmthoma!
